### PR TITLE
[BEAM-362] Move state-oriented remnants of OldDoFn to runners/core-java

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This moves `WindowingInternals`, `StateInternals`, and `TimerInternals` out of the Java SDK and into runners/core-java. This is motivated by code health and user-facing API health but also easy first steps towards easing alterations to `WatermarkHoldState`'s relationship with `OutputTimeFn`, which is going away.

Kudos to @jkff for moving `OldDoFn` which made all of this trivial.

This will require a Dataflow worker image.

### CODE REVIEW

**Comment by kennknowles** [comment]
(needs checkstyle touch ups)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/UnionCoderTest.java [hunk 1]
```diff
@@ -39,9 +37,4 @@ public void testCoderIsSerializable() {
     CoderProperties.coderSerializable(UnionCoder.of(ImmutableList.<Coder<?>>of(
         StringUtf8Coder.of(), DoubleCoder.of())));
   }
-
-  @Test
-  public void testCoderIsSerializableWithWellKnownCoderType() {
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/UnionCoderTest.java]: This seems unrelated to the current commit.

Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/join/UnionCoderTest.java]: The test is some kind of copy/paste error, not actually testing the class in question at all, that would make the SDK depend on runners/core.

**Comment by kennknowles** [comment]
R: @tgroh pushed new worker image, expect to see green

**Comment by kennknowles** [comment]
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Dataflow/2195/

**Comment by tgroh** [comment]
I'm happy with this (LGTM); any idea why Jenkins `Harness` seems to be misbehaving?

**Comment by kennknowles** [comment]
Unsure. I'll investigate and be sure there's no build breakage before I merge. If it is a real failure, it was introduced over the weekend. This PR was green on Friday and resolving conflicts was just another PR touching its imports.

**Comment by kennknowles** [comment]
retest this please
